# Book Recommendation System - EDA/Cleaning

The data comes from Kaggle's "Amazon Books Reviews" dataset: https://www.kaggle.com/datasets/mohamedbakhet/amazon-books-reviews/data?select=books_data.csv

The goal: create a book recommender system using sentiments on the reviews/review score to do so. Likely a content based one.

First, import necessary packages, then load in data.

In [1]:
# starting with basics, add more as needed

import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt
%matplotlib inline

In [2]:
ratings = pd.read_csv('../Data/Books_rating.csv')
books_data = pd.read_csv('../Data/books_data.csv')

Quick visualization to make sure everything loaded properly. 

In [3]:
ratings.head()

Id                           Title  Price         User_id  \
0  1882931173  Its Only Art If Its Well Hung!    NaN   AVCGYZL8FQQTD   
1  0826414346        Dr. Seuss: American Icon    NaN  A30TK6U7DNS82R   
2  0826414346        Dr. Seuss: American Icon    NaN  A3UH4UZ4RSVO82   
3  0826414346        Dr. Seuss: American Icon    NaN  A2MVUWT453QH61   
4  0826414346        Dr. Seuss: American Icon    NaN  A22X4XUPKF66MR   

                          profileName review/helpfulness  review/score  \
0               Jim of Oz "jim-of-oz"                7/7           4.0   
1                       Kevin Killian              10/10           5.0   
2                        John Granger              10/11           5.0   
3  Roy E. Perry "amateur philosopher"                7/7           4.0   
4     D. H. Richards "ninthwavestore"                3/3           4.0   

   review/time                                   review/summary  \
0    940636800           Nice collection of Julie Strain images   
1   1095724800                                Really Enjoyed It   
2   1078790400  Essential for every personal and Public Library   
3   1090713600  Phlip Nel gives silly Seuss a serious treatment   
4   1107993600                           Good academic overview   

                                         review/text  
0  This is only for Julie Strain fans. It's a col...  
1  I don't care much for Dr. Seuss but after read...  
2  If people become the books they read and if "t...  
3  Theodore Seuss Geisel (1904-1991), aka &quot;D...  
4  Philip Nel - Dr. Seuss: American IconThis is b...

In [4]:
books_data.head()

Title  \
0                     Its Only Art If Its Well Hung!   
1                           Dr. Seuss: American Icon   
2              Wonderful Worship in Smaller Churches   
3                      Whispers of the Wicked Saints   
4  Nation Dance: Religion, Identity and Cultural ...   

                                         description              authors  \
0                                                NaN     ['Julie Strain']   
1  Philip Nel takes a fascinating look into the k...       ['Philip Nel']   
2  This resource includes twelve principles in un...     ['David R. Ray']   
3  Julia Thomas finds her life spinning out of co...  ['Veronica Haddon']   
4                                                NaN      ['Edward Long']   

                                               image  \
0  http://books.google.com/books/content?id=DykPA...   
1  http://books.google.com/books/content?id=IjvHQ...   
2  http://books.google.com/books/content?id=2tsDA...   
3  http://books.google.com/books/content?id=aRSIg...   
4                                                NaN   

                                         previewLink  publisher publishedDate  \
0  http://books.google.nl/books?id=DykPAAAACAAJ&d...        NaN          1996   
1  http://books.google.nl/books?id=IjvHQsCn_pgC&p...  A&C Black    2005-01-01   
2  http://books.google.nl/books?id=2tsDAAAACAAJ&d...        NaN          2000   
3  http://books.google.nl/books?id=aRSIgJlq6JwC&d...  iUniverse       2005-02   
4  http://books.google.nl/books?id=399SPgAACAAJ&d...        NaN    2003-03-01   

                                            infoLink  \
0  http://books.google.nl/books?id=DykPAAAACAAJ&d...   
1  http://books.google.nl/books?id=IjvHQsCn_pgC&d...   
2  http://books.google.nl/books?id=2tsDAAAACAAJ&d...   
3  http://books.google.nl/books?id=aRSIgJlq6JwC&d...   
4  http://books.google.nl/books?id=399SPgAACAAJ&d...   

                      categories  ratingsCount  
0    ['Comics & Graphic Novels']           NaN  
1  ['Biography & Autobiography']           NaN  
2                   ['Religion']           NaN  
3                    ['Fiction']           NaN  
4                            NaN           NaN

Once we explore/clean up the files, we can merge on title. Let's explore the ratings dataset first. 

In [5]:
ratings.head()

Id                           Title  Price         User_id  \
0  1882931173  Its Only Art If Its Well Hung!    NaN   AVCGYZL8FQQTD   
1  0826414346        Dr. Seuss: American Icon    NaN  A30TK6U7DNS82R   
2  0826414346        Dr. Seuss: American Icon    NaN  A3UH4UZ4RSVO82   
3  0826414346        Dr. Seuss: American Icon    NaN  A2MVUWT453QH61   
4  0826414346        Dr. Seuss: American Icon    NaN  A22X4XUPKF66MR   

                          profileName review/helpfulness  review/score  \
0               Jim of Oz "jim-of-oz"                7/7           4.0   
1                       Kevin Killian              10/10           5.0   
2                        John Granger              10/11           5.0   
3  Roy E. Perry "amateur philosopher"                7/7           4.0   
4     D. H. Richards "ninthwavestore"                3/3           4.0   

   review/time                                   review/summary  \
0    940636800           Nice collection of Julie Strain images   
1   1095724800                                Really Enjoyed It   
2   1078790400  Essential for every personal and Public Library   
3   1090713600  Phlip Nel gives silly Seuss a serious treatment   
4   1107993600                           Good academic overview   

                                         review/text  
0  This is only for Julie Strain fans. It's a col...  
1  I don't care much for Dr. Seuss but after read...  
2  If people become the books they read and if "t...  
3  Theodore Seuss Geisel (1904-1991), aka &quot;D...  
4  Philip Nel - Dr. Seuss: American IconThis is b...

In [6]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000000 entries, 0 to 2999999
Data columns (total 10 columns):
 #   Column              Dtype  
---  ------              -----  
 0   Id                  object 
 1   Title               object 
 2   Price               float64
 3   User_id             object 
 4   profileName         object 
 5   review/helpfulness  object 
 6   review/score        float64
 7   review/time         int64  
 8   review/summary      object 
 9   review/text         object 
dtypes: float64(2), int64(1), object(7)
memory usage: 228.9+ MB


There are 3M row, 10 columns total. Let's check out missing values. 

In [7]:
ratings.isna().sum()

Id                          0
Title                     208
Price                 2518829
User_id                561787
profileName            561886
review/helpfulness          0
review/score                0
review/time                 0
review/summary             38
review/text                 8
dtype: int64

There are a lot of missing values for price, user_id, and profileName. Price is missing 2.5M of the 3M rows and can likely just be dropped. For User_id and profileName, we might be able to use those columns to fill in some missing values. 

Let's look at price and see if it can be salvaged at all. 

In [8]:
ratings['Price'].value_counts()

7.99      14729
14.95      8465
19.95      6341
24.95      6249
29.95      5612
          ...  
126.29        1
127.04        1
30.35         1
114.96        1
85.65         1
Name: Price, Length: 6004, dtype: int64

In [9]:
ratings['Price'].isna().mean()
# we're missing 84% of the data

0.8396096666666667

In [10]:
ratings['Price'].describe()
# both max and min price seem odd, likely typos or inaccurate.

count    481171.000000
mean         21.762656
std          26.206541
min           1.000000
25%          10.780000
50%          14.930000
75%          23.950000
max         995.000000
Name: Price, dtype: float64

In [11]:
ratings[ratings['Price'] == 995.0]

Id                                              Title  Price  \
1501750  0763706418  Visual Genetics Plus: Tutorial and Lab Simulat...  995.0   

               User_id          profileName review/helpfulness  review/score  \
1501750  A1PU04G5O644V  Christopher C. Lund                0/0           3.0   

         review/time                        review/summary  \
1501750    962323200  A Modest Treatise on Visual Genetics   

                                               review/text  
1501750  This is modest, yet suitable, book for an intr...

In [12]:
ratings[ratings['Price'] == 1.0].head()

# quick google search of Observational Before-After Studies in Road Safety shows the price is not $1, more like $90
# unless it was bought secondhand, which we have no real way of knowing, probably a typo

Id                                              Title  Price  \
37871   B000FO7L5S  Observational Before-After Studies in Road Safety    1.0   
37872   B000FO7L5S  Observational Before-After Studies in Road Safety    1.0   
100133  048640238X                                     Angel Notebook    1.0   
100134  048640238X                                     Angel Notebook    1.0   
170020  B000GPP42I  The Essential Australian Shepherd (Essential (...    1.0   

               User_id               profileName review/helpfulness  \
37871   A25H0OTASYWGSN        Anonymous Reviewer                3/3   
37872   A3ACIMBRKAVSW6  Isis T. Cabral "Eduardo"                0/1   
100133  A3MWRTCIUURO1H  S. McCoach "Love Angels"              15/16   
100134   AMDEYHRYEU06E                   N. Wood                1/1   
170020  A3GYF6RO0N9VFX             Tom McKiernan              15/18   

        review/score  review/time  \
37871            5.0   1105401600   
37872            5.0   1222646400   
100133           1.0   1095638400   
100134           1.0   1176163200   
170020           1.0   1009756800   

                                          review/summary  \
37871   Must-reading for those conducting safety studies   
37872                                 Greate negotiation   
100133                     It's a notebook - not a story   
100134                                           rip-off   
170020     Essentially I have read all of this elsewhere   

                                              review/text  
37871   This is one of the best technical books I've e...  
37872   The product arrived on time, and it fullfiled ...  
100133  I am not sure what reviewer before me is writi...  
100134  Amazon charges 2.99 for this, and it says righ...  
170020  This is about the best book on the subject I h...

Since we're missing 84% of the data, let's drop the column. 

In [13]:
ratings.drop(labels='Price', axis=1, inplace=True)
ratings.head()

Id                           Title         User_id  \
0  1882931173  Its Only Art If Its Well Hung!   AVCGYZL8FQQTD   
1  0826414346        Dr. Seuss: American Icon  A30TK6U7DNS82R   
2  0826414346        Dr. Seuss: American Icon  A3UH4UZ4RSVO82   
3  0826414346        Dr. Seuss: American Icon  A2MVUWT453QH61   
4  0826414346        Dr. Seuss: American Icon  A22X4XUPKF66MR   

                          profileName review/helpfulness  review/score  \
0               Jim of Oz "jim-of-oz"                7/7           4.0   
1                       Kevin Killian              10/10           5.0   
2                        John Granger              10/11           5.0   
3  Roy E. Perry "amateur philosopher"                7/7           4.0   
4     D. H. Richards "ninthwavestore"                3/3           4.0   

   review/time                                   review/summary  \
0    940636800           Nice collection of Julie Strain images   
1   1095724800                                Really Enjoyed It   
2   1078790400  Essential for every personal and Public Library   
3   1090713600  Phlip Nel gives silly Seuss a serious treatment   
4   1107993600                           Good academic overview   

                                         review/text  
0  This is only for Julie Strain fans. It's a col...  
1  I don't care much for Dr. Seuss but after read...  
2  If people become the books they read and if "t...  
3  Theodore Seuss Geisel (1904-1991), aka &quot;D...  
4  Philip Nel - Dr. Seuss: American IconThis is b...

In [14]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000000 entries, 0 to 2999999
Data columns (total 9 columns):
 #   Column              Dtype  
---  ------              -----  
 0   Id                  object 
 1   Title               object 
 2   User_id             object 
 3   profileName         object 
 4   review/helpfulness  object 
 5   review/score        float64
 6   review/time         int64  
 7   review/summary      object 
 8   review/text         object 
dtypes: float64(1), int64(1), object(7)
memory usage: 206.0+ MB


Next, let's see how many of these identifying values are unique. Let's look at id, title, user_id, and profileName. 

In [15]:
columns = ['Id', 'Title', 'User_id', 'profileName']

for column in columns:
    if column in ratings.columns:
        total_count = ratings[column].count()
        unique_count = ratings[column].nunique()
        print(f'Number of {column}: {total_count}')
        print(f'Number of unique {column}: {unique_count}\n')

Number of Id: 3000000
Number of unique Id: 221998

Number of Title: 2999792
Number of unique Title: 212403

Number of User_id: 2438213
Number of unique User_id: 1008972

Number of profileName: 2438114
Number of unique profileName: 854146



We can see that although there are 3M rows, there are only 221,998 unique book IDs. Similar metrics with title (keep in mind this does not account for any spelling/grammar differences in titles - not cleaned yet!), i.e., mutiple reviews per book. 

Let's look at review/time - it should be a timestamp, but it's an integer. Let's try converting to datetime. 

In [16]:
ratings['review/time'] = pd.to_datetime(ratings['review/time'])

In [17]:
ratings['review/time'].value_counts()

1970-01-01 00:00:01.346889600    5232
1970-01-01 00:00:01.361145600    4357
1970-01-01 00:00:01.355875200    4011
1970-01-01 00:00:01.357257600    3948
1970-01-01 00:00:01.168300800    3883
                                 ... 
1970-01-01 00:00:00.816134400       1
1970-01-01 00:00:00.824083200       1
1970-01-01 00:00:00.818208000       1
1970-01-01 00:00:00.815184000       1
1970-01-01 00:00:00.818035200       1
Name: review/time, Length: 6272, dtype: int64

Looks like everything is 1/1/1970 - deleting the column.

In [18]:
ratings.drop(labels='review/time', axis=1, inplace=True)
ratings.head()

Id                           Title         User_id  \
0  1882931173  Its Only Art If Its Well Hung!   AVCGYZL8FQQTD   
1  0826414346        Dr. Seuss: American Icon  A30TK6U7DNS82R   
2  0826414346        Dr. Seuss: American Icon  A3UH4UZ4RSVO82   
3  0826414346        Dr. Seuss: American Icon  A2MVUWT453QH61   
4  0826414346        Dr. Seuss: American Icon  A22X4XUPKF66MR   

                          profileName review/helpfulness  review/score  \
0               Jim of Oz "jim-of-oz"                7/7           4.0   
1                       Kevin Killian              10/10           5.0   
2                        John Granger              10/11           5.0   
3  Roy E. Perry "amateur philosopher"                7/7           4.0   
4     D. H. Richards "ninthwavestore"                3/3           4.0   

                                    review/summary  \
0           Nice collection of Julie Strain images   
1                                Really Enjoyed It   
2  Essential for every personal and Public Library   
3  Phlip Nel gives silly Seuss a serious treatment   
4                           Good academic overview   

                                         review/text  
0  This is only for Julie Strain fans. It's a col...  
1  I don't care much for Dr. Seuss but after read...  
2  If people become the books they read and if "t...  
3  Theodore Seuss Geisel (1904-1991), aka &quot;D...  
4  Philip Nel - Dr. Seuss: American IconThis is b...

Let's look at review/score next and see the range. 

In [19]:
ratings['review/score'].value_counts()
# seems like 1-5

5.0    1807343
4.0     585616
3.0     254295
1.0     201688
2.0     151058
Name: review/score, dtype: int64

In [20]:
ratings['review/score'].isna().sum()
# no missing values

0

Everything seems okay here. Let's look at review/helpfulness.

In [24]:
ratings['review/helpfulness'].value_counts()

0/0        885732
1/1        313000
2/2        156303
0/1        127865
1/2         99244
            ...  
62/102          1
76/120          1
173/230         1
23/118          1
62/111          1
Name: review/helpfulness, Length: 12084, dtype: int64

There are a lot of different scales here, such as 0/0, 1/1, 2/2, etc. Deleting the column.

In [25]:
ratings.drop(labels='review/helpfulness', axis=1, inplace=True)
ratings.head()

Id                           Title         User_id  \
0  1882931173  Its Only Art If Its Well Hung!   AVCGYZL8FQQTD   
1  0826414346        Dr. Seuss: American Icon  A30TK6U7DNS82R   
2  0826414346        Dr. Seuss: American Icon  A3UH4UZ4RSVO82   
3  0826414346        Dr. Seuss: American Icon  A2MVUWT453QH61   
4  0826414346        Dr. Seuss: American Icon  A22X4XUPKF66MR   

                          profileName  review/score  \
0               Jim of Oz "jim-of-oz"           4.0   
1                       Kevin Killian           5.0   
2                        John Granger           5.0   
3  Roy E. Perry "amateur philosopher"           4.0   
4     D. H. Richards "ninthwavestore"           4.0   

                                    review/summary  \
0           Nice collection of Julie Strain images   
1                                Really Enjoyed It   
2  Essential for every personal and Public Library   
3  Phlip Nel gives silly Seuss a serious treatment   
4                           Good academic overview   

                                         review/text  
0  This is only for Julie Strain fans. It's a col...  
1  I don't care much for Dr. Seuss but after read...  
2  If people become the books they read and if "t...  
3  Theodore Seuss Geisel (1904-1991), aka &quot;D...  
4  Philip Nel - Dr. Seuss: American IconThis is b...

Need to make sure title, review/summary, and review/text are cleaned. Looking at Title first. 

ratings:  
- explore userid/profilename relationship, see if it's useful
- make sure id matches book title
- make sure titles all match each other
- clean up review summary/text
- look for duplicates

books_data: 
- clean up titles/desriptions/authors
- remove links
- clean up publisher/date published

In [18]:
ratings.head()

Id                           Title         User_id  \
0  1882931173  Its Only Art If Its Well Hung!   AVCGYZL8FQQTD   
1  0826414346        Dr. Seuss: American Icon  A30TK6U7DNS82R   
2  0826414346        Dr. Seuss: American Icon  A3UH4UZ4RSVO82   
3  0826414346        Dr. Seuss: American Icon  A2MVUWT453QH61   
4  0826414346        Dr. Seuss: American Icon  A22X4XUPKF66MR   

                          profileName review/helpfulness  review/score  \
0               Jim of Oz "jim-of-oz"                7/7           4.0   
1                       Kevin Killian              10/10           5.0   
2                        John Granger              10/11           5.0   
3  Roy E. Perry "amateur philosopher"                7/7           4.0   
4     D. H. Richards "ninthwavestore"                3/3           4.0   

   review/time                                   review/summary  \
0    940636800           Nice collection of Julie Strain images   
1   1095724800                                Really Enjoyed It   
2   1078790400  Essential for every personal and Public Library   
3   1090713600  Phlip Nel gives silly Seuss a serious treatment   
4   1107993600                           Good academic overview   

                                         review/text  
0  This is only for Julie Strain fans. It's a col...  
1  I don't care much for Dr. Seuss but after read...  
2  If people become the books they read and if "t...  
3  Theodore Seuss Geisel (1904-1991), aka &quot;D...  
4  Philip Nel - Dr. Seuss: American IconThis is b...

Looking at books dataset next. 

In [26]:
books_data.head()

Title  \
0                     Its Only Art If Its Well Hung!   
1                           Dr. Seuss: American Icon   
2              Wonderful Worship in Smaller Churches   
3                      Whispers of the Wicked Saints   
4  Nation Dance: Religion, Identity and Cultural ...   

                                         description              authors  \
0                                                NaN     ['Julie Strain']   
1  Philip Nel takes a fascinating look into the k...       ['Philip Nel']   
2  This resource includes twelve principles in un...     ['David R. Ray']   
3  Julia Thomas finds her life spinning out of co...  ['Veronica Haddon']   
4                                                NaN      ['Edward Long']   

                                               image  \
0  http://books.google.com/books/content?id=DykPA...   
1  http://books.google.com/books/content?id=IjvHQ...   
2  http://books.google.com/books/content?id=2tsDA...   
3  http://books.google.com/books/content?id=aRSIg...   
4                                                NaN   

                                         previewLink  publisher publishedDate  \
0  http://books.google.nl/books?id=DykPAAAACAAJ&d...        NaN          1996   
1  http://books.google.nl/books?id=IjvHQsCn_pgC&p...  A&C Black    2005-01-01   
2  http://books.google.nl/books?id=2tsDAAAACAAJ&d...        NaN          2000   
3  http://books.google.nl/books?id=aRSIgJlq6JwC&d...  iUniverse       2005-02   
4  http://books.google.nl/books?id=399SPgAACAAJ&d...        NaN    2003-03-01   

                                            infoLink  \
0  http://books.google.nl/books?id=DykPAAAACAAJ&d...   
1  http://books.google.nl/books?id=IjvHQsCn_pgC&d...   
2  http://books.google.nl/books?id=2tsDAAAACAAJ&d...   
3  http://books.google.nl/books?id=aRSIgJlq6JwC&d...   
4  http://books.google.nl/books?id=399SPgAACAAJ&d...   

                      categories  ratingsCount  
0    ['Comics & Graphic Novels']           NaN  
1  ['Biography & Autobiography']           NaN  
2                   ['Religion']           NaN  
3                    ['Fiction']           NaN  
4                            NaN           NaN

Let's get rid of links, we won't use them in analysis. 

In [27]:
books_data.drop(labels=['image', 'previewLink', 'infoLink'], axis=1, inplace=True)
books_data.head()

Title  \
0                     Its Only Art If Its Well Hung!   
1                           Dr. Seuss: American Icon   
2              Wonderful Worship in Smaller Churches   
3                      Whispers of the Wicked Saints   
4  Nation Dance: Religion, Identity and Cultural ...   

                                         description              authors  \
0                                                NaN     ['Julie Strain']   
1  Philip Nel takes a fascinating look into the k...       ['Philip Nel']   
2  This resource includes twelve principles in un...     ['David R. Ray']   
3  Julia Thomas finds her life spinning out of co...  ['Veronica Haddon']   
4                                                NaN      ['Edward Long']   

   publisher publishedDate                     categories  ratingsCount  
0        NaN          1996    ['Comics & Graphic Novels']           NaN  
1  A&C Black    2005-01-01  ['Biography & Autobiography']           NaN  
2        NaN          2000                   ['Religion']           NaN  
3  iUniverse       2005-02                    ['Fiction']           NaN  
4        NaN    2003-03-01                            NaN           NaN

We'll need to make sure title, authors, categories are all cleaned (text-wise). Let's look for missing values. 

In [28]:
books_data.isna().sum()

Title                 1
description       68442
authors           31413
publisher         75886
publishedDate     25305
categories        41199
ratingsCount     162652
dtype: int64

In [29]:
books_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 212404 entries, 0 to 212403
Data columns (total 7 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Title          212403 non-null  object 
 1   description    143962 non-null  object 
 2   authors        180991 non-null  object 
 3   publisher      136518 non-null  object 
 4   publishedDate  187099 non-null  object 
 5   categories     171205 non-null  object 
 6   ratingsCount   49752 non-null   float64
dtypes: float64(1), object(6)
memory usage: 11.3+ MB


About half of ratingsCounts is missing, and there are a decent number of missing values for the rest of the columns except for title. 

Let's look at ratingsCount. 

In [30]:
books_data['ratingsCount'].value_counts()

1.0       20919
2.0        8249
3.0        4393
4.0        2798
5.0        1995
          ...  
321.0         1
2569.0        1
221.0         1
3596.0        1
1779.0        1
Name: ratingsCount, Length: 478, dtype: int64

Merge ratings and books data on title. 